qiskit/synthesis/boolean/boolean_expression.py

```diff
@@ -0,0 +1,223 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""A class for parsing and synthesizing boolean expressions"""
+
+import ast
+import itertools
+import re
+from os.path import isfile
+from typing import Union
+import numpy as np
+
+from .boolean_expression_visitor import (
+    BooleanExpressionEvalVisitor,
+    BooleanExpressionArgsCollectorVisitor,
+)
+
+
+class TruthTable:
+    """A simple implementation of a truth table for a boolean function"""
+
+    EXPLICIT_REP_THRESHOLD = (
+        12  # above this number of bits, do not explicitly save the values in a list
+    )
+
+    def __init__(self, func, num_bits):
+        self.func = func
+        self.num_bits = num_bits
+        self.explicit_storage = self.num_bits <= self.EXPLICIT_REP_THRESHOLD
+        if self.explicit_storage:
+            self.values = np.array(
+                [self.func(assignment) for assignment in self.all_assignments()],
+                dtype=bool,
+            )
+        else:
+            self.implicit_values = {}
+
+    def all_assignments(self):
+        """Return an ordered list of all assignments, ordered from left to right
+        i.e. 000, 100, 010, 110, 001, 101, 011, 111"""
+        return [
+            tuple(reversed(assignment))
+            for assignment in itertools.product([False, True], repeat=self.num_bits)
+        ]
+
+    def __getitem__(self, key):
+        if isinstance(key, str):
+            key = (bit != "0" for bit in key)
+        if self.explicit_storage:
+            key = sum(2**n for n, bit in enumerate(key) if bit)
+            return self.values[key]
+        else:
+            return self.implicit_values.setdefault(key, self.func(key))
```

Comment:
Why do we need to insert `(key, value)` to `implicit_values`? I can see how this can be an optimization when we need the same value multiple times, but if the calling algorithm goes over all assignments, wouldn't this build the full truth table?

Reply:
The `TruthTable` object is created to be somewhat generic, without a specific algorithm in mind. If we switch to an algorithm that doesn't have to read the whole truth table for whatever reason, I still want it to function correctly and with relative efficiency (i.e. never use `func` twice on the same input if possible).

Reply:
In this case, I would suggest to reduce the size of the cache if it gets too large (for example, simply remove all cached values).